Rewrite instruction (#415)
neuefische/web-exercises: sessions/recap-project-5/art-gallery-app/README.md

🚀 Project Deployment to Vercel is required: Please follow [these instructions](https://github.com/neuefische/web-protocol-template/blob/main/docs/deployment-vercel.md) to deploy the project on Vercel.

## Tasks

In this app, users will have access to a **Spotlight**, **Gallery**, **Favorite**, and **Detail** page.

### 1. Spotlight Page

This page will serve as the app's homepage. A random art piece image, along with its artist’s name, will be displayed as a spotlight feature.

#### Value Proposition

**As an** art enthusiast

**I want to** see a spotlight art piece

**so that** I can get inspirational highlights.

#### Acceptance Criteria

- [ ] One art piece is randomly selected as a spotlight feature.
- [ ] The image of the selected art piece is displayed.
- [ ] The artist's name for the spotlighted art piece is displayed.

#### Tasks

- [ ] Write a function to select a random art piece. [Refer to this resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#examples).
- [ ] Create a new component for the Spotlight page.

> - [ ] Write tests for the Spotlight component.

Reviewer: do we really want to add tests in this recap project? It makes sense since they learn about it prior, but maybe its a bit too much and they get stuck?

Author: I wrote with all the test that we have at SP, but we could select certain test for them to do. Select certain components.

Reviewer: What do you mean exactly, I didn't understand 😅

- [ ] Clicking "Spotlight" shows the Spotlight Page.
- [ ] Clicking "Gallery" shows the Gallery Page.

#### Tasks

- [ ] Create a Navigation component that renders all navigation links.
- [ ] Render the Navigation component only once, ensuring it appears on all pages.
- [ ] Write tests for the Navigation component to confirm that the links are rendered correctly.

### 4. Art Piece Details Page

When a user clicks on an art piece, they should see all detailed information about it, along with an option to navigate back to the list of art pieces.

#### Value Proposition

**As an** art enthusiast

**I want to** view detailed information about an art piece

**so that** I can learn everything about it.

#### Acceptance Criteria

- [ ] Clicking on an art piece from the list navigates to the detail page.
- [ ] The art piece’s image is displayed.
- [ ] The art piece’s title is displayed.
- [ ] The art piece’s artist is displayed.
- [ ] The art piece’s year is displayed.
- [ ] The art piece’s genre is displayed.
- [ ] A back button is displayed.
- [ ] Clicking the back button navigates back to the list view.

#### Tasks

- [ ] Create a new component for displaying Art Piece Details.

> - [ ] Create a single page that renders the Art Piece Details, displaying information for the selected art piece.

Reviewer: ```suggestion
- [ ] Create a single dynamic page that renders the Art Piece Details, displaying information for the selected art piece.
```

- [ ] Write tests for the Art Piece Details component to confirm that all information is rendered correctly.

### 5. Favorite Page

The Favorite page will display a list of all favorited art pieces. Each art piece will show its image, artist’s name, title, and a favorite button.

Users should be able to save their favorite art pieces both on the Gallery page and the Detail page.

To achieve this, we’ll divide the tasks into two parts:

1. Implement the favorite button where needed.
2. Create the Favorite page.

### 5.1. Favorite Button

#### Value Proposition

**As an** art enthusiast

**I want to** mark art pieces as favorites

**so that** I can find them more easily.

#### Acceptance Criteria

- [ ] The favorite button is displayed on each entry in the list view.
- [ ] The favorite button is displayed in the details view.

> - [ ] The favorite button is **NOT** displayed in the spotlight view.

Reviewer: why not? I think it makes sense to keep the favorite button there as well.

Author: It was like that for NF. I didn't change it.
If you want we can change it, that the button to `The favorite button is displayed in the spotlight view.`

Reviewer: yeah lets change that, it makes sense from a UX perspective and students can reuse the ArtpiecePreview component.

- [ ] Clicking the favorite button on a non-favorite piece saves it as a favorite.

> - [ ] Clicking the favorite button on a favorited piece removes it from favorites.

Reviewer: ```suggestion
- [ ] Clicking the favorite button on a favorited piece removes it from favorites.
- [ ] The favorite state of an art piece is visually highlighted by changing the color of the favorite button. 
```

#### Tasks

- [ ] Create a new component for the favorite button.

> - [ ] Store this additional information (favorite status) in a state. We will also be adding comments to each art piece, so consider how to store both types of information together.

Author: Thank you for the update! Could you clarify when the decision to split the state was made? In the NF instructions, it was still a single state as `artPiecesInfo`, which I believe has benefits worth considering.

Typically, it’s best practice to keep related information consolidated in a single state, as this often leads to a more maintainable and scalable codebase. By keeping the information together, students can get accustomed to managing complex state structures, which can be valuable for more advanced projects.

> - [ ] Store this additional information (favorite status) in a state. We will also be adding comments to each art piece, so consider how to store both types of information together.

Author: We decided to add some example for both way and let the student decide which to use.

> - [ ] Write tests to verify that the favorite button appears in the list view and details view but is not shown in the spotlight.

Reviewer: I think we should give the students a bit more guidance here on how to check whether a certain art piece is favorite or not. Maybe something like "Create a function for checking whether an art Piece is included in the favoriteIds list".

Author: Added it as a hint.

- [ ] Each art piece's artist name is displayed.
- [ ] Each art piece has an active favorite button.

#### Tasks

- [ ] Reuse an existing component to display only the favorite art pieces.
- [ ] Write tests to verify that the navigation link labeled "Favorites" is displayed in all pages.

### 6. Comments for Art Pieces

To enhance user engagement, we’ll add a commenting feature that allows users to write and view comments on each art piece. A comment section will be displayed on the details page, showing comments specifically for the art piece being viewed.

#### Value Proposition

**As an** art enthusiast

**I want to** write comments on art pieces

**so that** I can note down my thoughts and ideas about the work.

#### Acceptance Criteria

- [ ] The detail view includes a section labeled "Comments" displaying a list of comments for this art piece.
- [ ] Each comment’s text is displayed.
- [ ] Each comment’s date and time are displayed.
- [ ] The detail view includes an input field for writing a comment.
- [ ] The detail view includes a submit button labeled "Send".
- [ ] After submitting the form, the new comment is added to the list of comments.

#### Tasks

> - [ ] Create at least two new components for managing and displaying comments.

Reviewer: Maybe we can be a bit more explicit here as well. Especially on how to store the comments in a global state.

Author: Added it as a hint.

🚀 Project Deployment to Vercel is required: Please follow [these instructions](https://github.com/neuefische/web-protocol-template/blob/main/docs/deployment-vercel.md) to deploy the project on Vercel.

## Tasks

This app will include four main pages: **Spotlight**, **Gallery**, **Favorite**, and **Detail**.

Each page will need access to the information from the API. Consider where to fetch and store the art piece data to ensure that all pages can easily access it.

<details>
<summary>Hints 🙈</summary>
In Next.js, the place to store data or states that need to be accessed across the app is in `_app.js`.

It's best practice to consolidate related information in a single state. However, if you're not comfortable with this approach, you can use separate states. Keep in mind that you’ll need to identify which information corresponds to each art piece, so consider saving the slug or a unique identifier in each state to maintain clarity.

</details>

> ### 1. Spotlight Page

Author: Thanks for the suggestion! I think I understand your perspective, but I’d like to clarify a couple of points to ensure we’re aligned.

Are you suggesting that we initially have students fetch the data on the Gallery page and then move it to `__app`? Personally, I feel that having them start with fetching data in one place and then moving it later could lead to unnecessary complexity without adding much learning value. I think it’s beneficial for students to first think critically about *where* data should be fetched based on app needs rather than having them rework that setup.

However, if you’re simply suggesting that students start with building the Gallery page first, followed by the Spotlight page, I’m completely on board with that! Reordering the tasks to start with the Gallery would work well without requiring them to rework data fetching.

One thought: if we do want them to practice changing the data fetch location as a learning exercise, maybe we could create a dedicated challenge in the data-fetching session. This way, they could focus on it as a standalone concept.

Additionally, all tasks are numbered, so if it helps with clarity, we could add “User Story #” before each task to reinforce the order.

Let me know what you think!

- [ ] A navigation link labeled "Gallery" is displayed.
- [ ] Clicking "Spotlight" shows the Spotlight Page.
- [ ] Clicking "Gallery" shows the Gallery Page.

#### Tasks

- [ ] Create a Navigation component that renders all navigation links.
- [ ] Render the Navigation component only once, ensuring it appears on all pages.

### 4. Art Piece Details Page

When a user clicks on an art piece, they should see all detailed information about it, along with an option to navigate back to the list of art pieces.

#### Value Proposition

**As an** art enthusiast

**I want to** view detailed information about an art piece

**so that** I can learn everything about it.

#### Acceptance Criteria

- [ ] Clicking on an art piece from the list navigates to the detail page.
- [ ] The art piece’s image is displayed.
- [ ] The art piece’s title is displayed.
- [ ] The art piece’s artist is displayed.
- [ ] The art piece’s year is displayed.
- [ ] The art piece’s genre is displayed.
- [ ] A back button is displayed.
- [ ] Clicking the back button navigates back to the list view.

#### Tasks

- [ ] Create a new component for displaying Art Piece Details.

> - [ ] Create a single dynamic page that renders the Art Piece Details, displaying information for the selected art piece.

Author: I bold the **dynamic page** to make it more clear.

### 5. Favorite Page

The Favorite page will display a list of all favorite art pieces. Each art piece will show its image, artist’s name, title, and a favorite button.

Users should be able to save their favorite art pieces both on the Gallery page and the Detail page.

To achieve this, we’ll divide the tasks into two parts:

1. Implement the favorite button where needed.
2. Create the Favorite page.

> ### 5.1. Favorite Button

Reviewer 2: In my opinion "Favorite Button" is not a descriptive feature name. While it's fine to split the "Bookmark" feature, I think it's important to choose feature-based names for user stories to set a good example.

 For example, a title like "Mark Art Pieces as Favorites" would better represent the feature not the UI element.

```suggestion
### 5.1. Mark Art Pieces as Favorites
```

**Example:**

- **Using a Single State:**

   ```js
   const [carInfo, setCarInfo] = useState([
      { model: "vw-taos", colors: ["red", "blue", "white"], automatic: true }, 
      { model: "vw-golf", colors: [ "black", "white"], automatic: false }
      ]);
   ```

- **Separate States per information:**

   ```js
   const [carColors, setCarColors] = useState([
      { model: "vw-taos", colors: "red"},
      { model: "vw-taos", colors: "blue" }, 
      { model: "vw-golf", colors: "black" }
      ]);
   
   const [carAutomatic, setCarAutomatic] = useState([
       "vw-taos", 
       "vw-golf",
      ]);
   ```

</details>

#### Value Proposition

**As an** art enthusiast

**I want to** mark art pieces as favorites

**so that** I can find them more easily.

#### Acceptance Criteria

- [ ] The favorite button is displayed on each entry in the list view.
- [ ] The favorite button is displayed in the details view.
- [ ] The favorite button is displayed in the spotlight view.

> - [ ] Clicking the favorite button on a non-favorite piece saves it as a favorite.
> - [ ] Clicking the favorite button on a favorite piece removes it from favorites.
> - [ ] The favorite state of an art piece is visually highlighted by changing the color of the favorite button.

Reviewer 2: At the risk of being super annoying....

I think these criteria could be improved because ssers don’t see the technical action of saving a favorite, they just see a visual change (like color or icon). 

Therefore I suggest to focus on the visible feedback.

```suggestion
- [ ] Clicking the favorite button on a non-favorite piece, it visibly indicates that the piece is now a favorite (e.g., by changing color or icon).
- [ ] Clicking the favorite button on a favorite piece, it visibly indicates that the piece is no longer a favorite.
- [ ] Favorite art pieces are visually highlighted, clearly showing their favorite status.
```

- [ ] Each comment’s text is displayed.
- [ ] Each comment’s date and time are displayed.
- [ ] The detail view includes an input field for writing a comment.
- [ ] The detail view includes a submit button labeled "Send".
- [ ] After submitting the form, the new comment is added to the list of comments.

#### Tasks

- [ ] Create at least two new components: one for managing the comment input and submission, and another for displaying the list of comments.
- [ ] Store the comment information in a state. You can either create a new state specifically for comments or add this information to an existing state, ensuring it’s easily accessible for each relevant art piece.

### 8. Persist Favorites and Comments in the Browser

Currently, all data is saved using `useState`, meaning it’s lost upon page refresh. To improve the user experience, we want to persist data in local storage. Keep in mind that when there is no data in local storage, we’ll need to fetch everything from the API.

#### Value Proposition

**As an** art enthusiast

**I want to** persist my favorites and comments

> **so that** my data remains saved even if I close the app.

Reviewer 2: I think we could make this more user-centered as well. For that reason, I suggest the following:

**As an** art enthusiast
**I want** my favorites and comments to be saved,
**so that** my data remains accessible even if I close or refresh the app.

## Tasks

This app will include four main pages: **Spotlight**, **Gallery**, **Favorite**, and **Detail**.

Each page will need access to the information from the API. Consider where to fetch and store the art piece data to ensure that all pages can easily access it.

<details>
<summary>Hint ✨</summary>
In Next.js, the place to store data or states that need to be accessed across the app is in <code>_app.js</code>.

> It's best practice to consolidate related information in a single state. However, if you're not comfortable with this approach, you can use separate states. Keep in mind that you’ll need to identify which information corresponds to each art piece, so consider saving the slug or a unique identifier in each state to maintain clarity.

Reviewer: I would keep the solution open without ranking them, this would lead to students to overwhelm themselves I guess. 
````suggestion
There are multiple ways to store the local information of your art pieces:
1. Single "info" state:
```js
const carInfo = useState([
  {carId: '1234' , isStarred: false, comments: []},
  {carId: '5678' , isStarred: true, comments: []},
])
```
2. Separate States per information:
```js
const starredCarIds = useState(["1234","5678"])
const comments = useState({
  id: '1',
  carId: '1234',
  comment: 'comment 1'
},{
    id: '2',
  carId: '1234',
  comment: 'comment 2'
}])
```
````

Author: Nice, I combine it with the one I did.
Let me know.
I move it of place to be near where the student may need this hint.

### 5.1. Favorite Button

Users want the ability to add additional information to each art piece. Consider where to store the art piece data to ensure all pages can access it efficiently. Remember that we’ll be reusing this favorite button across multiple pages.

<details>
<summary>Hint ✨</summary>

In Next.js, the recommended place to store data or states that need to be accessed across the app is in `_app.js`.

It’s best practice to consolidate related information into a single state. However, if you're not comfortable with this approach, you can use separate states. Just ensure you can clearly identify which information corresponds to each art piece—consider saving a unique identifier, such as a slug, within each state for clarity.

**Example:**

- **Using Multiple States:**

   ```js
   const [carColors, setCarColors] = useState({ model: "vw-taos", colors: ["red", "blue", "white"] });
   const [carAutomatic, setCarAutomatic] = useState({ model: "vw-taos", automatic: true });

>    ```

Reviewer: Ah well I think this version does not have the benefits of the multi state solution. Please checkout my comment.

Author: I combine them.

#### Tasks

- [ ] Reuse an existing component to display only the favorite art pieces.

<details>
<summary>Hint ✨</summary>
You could write a new function for checking whether an art Piece is favorite.

</details>

### 6. Testing Components

As developers, we aim to avoid repetitive manual testing by automating component tests.

#### Value Proposition

**As a** programmer

**I want to** automate component testing

**so that** I can identify bugs more easily.

#### Acceptance Criteria

- [ ] Automated tests cover all core components, ensuring they render and function as expected.

#### Tasks

- [ ] Write tests for the Spotlight component.
- [ ] Write tests for the Art Piece Preview and List of Art Pieces components.
- [ ] Write tests for the Navigation component to confirm that links are rendered correctly.
- [ ] Write tests for the Art Piece Details component to confirm that all details are displayed correctly.

> 

Reviewer: I think thats perfect.
